### PR TITLE
Productionnificationize

### PR DESCRIPTION
**Pull request by schwers**
Depends on improvements to @r/build from https://github.com/reddit/node-build/pull/19

### CODE REVIEW

**Comment by spladug** [comment]
Will modmail need similar tweaks?

**Comment by schwers** [comment]
👓 @phil303 or @nramadas

File: src/Client.js [hunk 1]
```diff
@@ -80,7 +80,7 @@ const client = Client({
     return data;
   },
   appComponent: <App/>,
-  debug: true,
+  debug: (process.env.NODE_ENV || 'production') !== 'production',
```
Comment (nramadas) [inline src/Client.js]: reminder to remove the `||`

File: blueprints.config.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+var configs = require('@r/build/lib/configs');
+
+module.exports = function(isProduction) {
+  var clientConfig = configs.getClientConfig(isProduction);
+  var serverConfig = configs.getServerConfig(isProduction);
+
+  // if we're in production mode, clean the build directory because its going
+  // to be copied to s3 in its entirety
+  if (isProduction) {
```
Comment (phil303) [inline blueprints.config.js]: should this be included in the generic prod client config?

File: blueprints.config.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    clientConfig.webpack.plugins = [{
+      generator: 'clean-directories',
+      paths: [ 'build/' ],
+    }].concat(clientConfig.webpack.plugins);
+  }
+
+  // Copy static files for deploying / serving. We also use these in debug
+  // to more closely mimic production.
+  serverConfig.webpack.plugins = serverConfig.webpack.plugins.concat([{
```
Comment (phil303) [inline blueprints.config.js]: same question here

Comment (phil303) [inline blueprints.config.js]: woops, deleted my comment but same question here

**Comment by phil303** [comment]
:nail_care:

File: blueprints.config.js (hunk 2 above)
Comment (schwers) [inline blueprints.config.js]: Yeah this could be in the default prod config, I'll move it.

File: blueprints.config.js (hunk 3 above)
Comment (schwers) [inline blueprints.config.js]: Modmail, and potentially other projects, have a slightly different assets folder. I thought it was better to have it stay explicit and per-project. I'd be open to moving it though

**Comment by schwers** [comment]
@spladug Yup, it shouldn't be too much work though. Which project is easier for you to test with?

**Comment by schwers** [comment]
💇

File: src/server/templates/main.jsx [hunk 4]
```diff
@@ -35,7 +39,13 @@ export default function(data, store) {
...
           } }
         />
-        <script type='text/javascript' src={ `${assetPath}/${clientJsSrc}` }></script>
+        <script
+          id='data'
+          dangerouslySetInnerHTML={ {
+            __html: `window.___r = ${JSON.stringify(data)}`,
+          } }
+        ></script>
+        <script type='text/javascript' src={ `${assetPath}/${JS_FILE}` } />
```
Comment (phil303) [inline src/server/templates/main.jsx]: just noticed this. Should this be `async`?

**Comment by phil303** [comment]
one last question, but :fish:-worthy

**Comment by spladug** [comment]
@schwers modmail's the one i'm supposed to be launching this week, so that'll need to happen shortly. i can test on this for now though. thanks!

File: src/server/templates/main.jsx (hunk 4 above)
Comment (schwers) [inline src/server/templates/main.jsx]: good catch

**Pull request merged**